gammapy/cube/tests/test_fit.py

```diff
@@ -77,75 +78,72 @@ def sky_model():
+    # define energy dispersion
+    e_true = geom_etrue.get_axis_by_name("energy").edges
+    e_reco = geom.get_axis_by_name("energy").edges
+    edisp = EnergyDispersion.from_diagonal_response(e_true=e_true, e_reco=e_reco)
+
+    # define fit mask
+    center = sky_model.spatial_model.position
+    circle = CircleSkyRegion(center=center, radius=1 * u.deg)
+    mask_fit = background_model.map.geom.region_mask([circle])
+
+    return MapDataset(
         model=sky_model,
         exposure=exposure,
-        background_model=background,
+        background_model=background_model,
         psf=psf,
         edisp=edisp,
-    ).npred()
-    return npred
+        mask_fit=mask_fit,
+        **kwargs
+    )
 
 
-@requires_dependency("iminuit")
 @requires_data()
-def test_map_fit(sky_model):
-    ebounds = np.logspace(-1.0, 1.0, 3)
-    ebounds_true = np.logspace(-1.0, 1.0, 4)
-    geom_r = geom(ebounds)
-    geom_t = geom_etrue(ebounds_true)
-
-    background_map = background(geom_r)
-    background_model_1 = BackgroundModel(background_map, norm=0.5)
-    background_model_2 = BackgroundModel(background_map, norm=1)
-
-    psf_map = psf(geom_t)
-    edisp_map = edisp(geom_r, geom_t)
-    exposure_map = exposure(geom_t)
-    counts_map_1 = counts(
-        sky_model, exposure_map, background_model_1, psf_map, edisp_map
-    )
-    counts_map_2 = counts(
-        sky_model, exposure_map, background_model_2, psf_map, edisp_map
-    )
+def test_map_dataset_fits_io(tmpdir, sky_model, geom, geom_etrue):
+    dataset = get_map_dataset(sky_model, geom , geom_etrue)
+    dataset.counts = dataset.npred()
+    dataset.mask_safe = dataset.mask_fit
 
-    mask_map = mask_fit(geom_r, sky_model)
-    sky_model.parameters["sigma"].frozen = True
+    dataset.write(tmpdir / "test.fits")
 
-    dataset_1 = MapDataset(
-        model=sky_model,
-        counts=counts_map_1,
-        exposure=exposure_map,
-        mask_fit=mask_map,
-        psf=psf_map,
-        edisp=edisp_map,
-        background_model=background_model_1,
-        evaluation_mode="local",
-    )
+    dataset_new = MapDataset.read(tmpdir / "test.fits")
 
-    dataset_2 = MapDataset(
-        model=sky_model,
-        counts=counts_map_2,
-        exposure=exposure_map,
-        mask_fit=mask_map,
-        psf=psf_map,
-        edisp=edisp_map,
-        background_model=background_model_2,
-        evaluation_mode="global",
-        likelihood="cstat",
-    )
+    assert dataset_new.model is None
+
+    assert_allclose(dataset.counts.data, dataset_new.counts.data)
+    assert_allclose(dataset.background_model.map.data, dataset_new.background_model.map.data)
+    assert_allclose(dataset.edisp.data.data.value, dataset_new.edisp.data.data.value)
+    assert_allclose(dataset.psf.data, dataset_new.psf.data)
+    assert_allclose(dataset.exposure.data, dataset_new.exposure.data)
+    assert_allclose(dataset.mask_fit, dataset_new.mask_fit)
+    assert_allclose(dataset.mask_safe, dataset_new.mask_safe)
```

Comment:
Extend this test a bit? Just asserting on the data, the following things could still go wrong and you won't notice:

- some HDU name changes and existing files become unreadable
- somehow the energy HDUs clobber or have non-unique names and thus aren't read back OK

So my suggestion is to add more asserts: on the HDU names in the FITS HDUList, and also asserts that check if the energy axes are read back OK.

Reply:
Done

Reply:
I'm pretty sure we will get issues in the coming months and years because we keep changing the content and HDU names for our FITS serialisations. Probably we'll have to introduce versions at some point, to be able to read back old files with newer Gammapy versions.

Would be really good to establish via a test what we store at the moment, i.e. to add an assert on what `[_.name for _ in hdulist]` is. Then we'll be more conscious of data file content changes, because the test fails. If you don't have that, it's easy to change whatever you like and things keep working, because you don't have a stable reference file.

Reply:
I adde the additional asserts on the hdu names.